src/CommandLine.cc

```diff
@@ -0,0 +1,163 @@
+    for (int i = 0; i < argC; ++i)
+        argv_.push_back(xstrdup(argV[i]));
+    argv_.push_back(nullptr); // POSIX argv "must be terminated by a null pointer"
+
+    /* copy grammar rules for the long options */
+    if (longRules) {
+        for (auto longOption = longRules; longOption->name; ++longOption)
+            longOptions_.emplace_back(*longOption);
+        longOptions_.emplace_back();
+    }
+}
+
+CommandLine::CommandLine(const CommandLine &them):
+    CommandLine(them.argc(), them.argv(), *them.shortOptions_, them.longOptions())
+{
+}
+
+CommandLine &
+CommandLine::operator =(const CommandLine &them)
+{
+    // cannot just swap(*this, them): std::swap(T,T) may call this assignment op
+    CommandLine tmp(them);
+    std::swap(argv_, tmp.argv_);
+    std::swap(shortOptions_, tmp.shortOptions_);
+    std::swap(longOptions_, tmp.longOptions_);
+    return *this;
+}
+
+CommandLine::~CommandLine()
+{
+    for (auto arg: argv_)
+        xfree(arg);
+
+    xfree(shortOptions_);
+}
+
+bool
+CommandLine::hasOption(const int optIdToFind, const char **optValue) const
+{
+    ResetGetopt(false); // avoid duped warnings; forEachOption() will complain
+    int optId = 0;
+    while (nextOption(optId)) {
+        if (optId == optIdToFind) {
+            if (optValue)
+                *optValue = optarg;
+            return true;
+        }
+    }
+    return false;
+}
+
+void
+CommandLine::forEachOption(Visitor visitor) const
+{
+    ResetGetopt(true);
+    int optId = 0;
+    while (nextOption(optId))
+        visitor(optId, optarg);
+}
+
+/// extracts the next option (if any)
+/// \returns whether the option was extracted
+/// throws on unknown option or missing required argument
+bool
+CommandLine::nextOption(int &optId) const
+{
+    optId = getopt_long(argc(), argv(), shortOptions_, longOptions(), nullptr);
+    if ((optId == ':' && shortOptions_[0] == ':') || optId == '?') {
+        assert(optind > 0 && static_cast<unsigned int>(optind) < argv_.size());
+        SBuf errMsg;
+        errMsg.Printf("'%s': %s", argv_[optind - 1],  optId == '?' ?
+                      "unrecognized option or missing required argument" : "missing required argument");
+        throw TexcHere(errMsg);
+    }
+    return optId != -1;
+}
+
+void
+CommandLine::resetArg0(const char *programName)
+{
+    assert(programName);
+    xfree(argv_[0]);
+    argv_[0] = xstrdup(programName);
+}
+
+void
+CommandLine::pushFrontOption(const char *name, const char *value)
+{
+    assert(name);
+    argv_.insert(argv_.begin() + 1, xstrdup(name));
+    if (value)
+        argv_.insert(argv_.begin() + 2, xstrdup(value));
+}
+
+Option::Option() :
+    option({nullptr, 0, nullptr, 0})
+{
+}
+
+Option::Option(const LongOption &opt)
+{
+    copy(opt);
+}
+
+Option::Option(const Option &opt):
+    Option(static_cast<const LongOption &>(opt))
+{
+}
+
+Option::~Option()
+{
+    xfree(name);
+}
+
+Option &
+Option::operator =(const Option &opt)
+{
+    xfree(name);
```

Comment:
Missing protection from assignment to self.

Reply:
Done.